File: Minimum_Window_Substring.cpp
/******************************************************************************

Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).

Example:

Input: S = "ADOBECODEBANC", T = "ABC"
Output: "BANC"

https://leetcode.com/problems/minimum-window-substring/

*******************************************************************************/

class Solution {
public:
    string minWindow(string s, string t) {


        unordered_map<char, int> letters;
        for(auto c : t) letters[c]++;
        int count = 0;
        int low = 0, min_length = INT_MAX, min_start = 0;
        for(int high = 0; high<s.length(); high++) {
            if(letters[s[high]] > 0) count++;
            letters[s[high]]--;
            if(count == t.length()) {
                while(low < high && letters[s[low]] < 0) letters[s[low]]++, low++;
                if(min_length > high-low) min_length = high-(min_start=low)+1;
                letters[s[low++]]++;
                count--;
            }
        }
        return min_length == INT_MAX ? "" : s.substr(min_start, min_length);

    }
};

File: 08. Hashing/Minimum_Window_Substring.java
/*
      Given 2 strings : 
      S = "TUTMTAPTAT";
      T = "TTA";
	
      Find the minimum Window in S that present all T String characters.
      return the minimum window substring.
*/

class Solution{
        public String minWindow(String s, String t) {
        	Map<Character,Integer> map = new HashMap<>();
        	for(char c: t.toCharArray())
            		map.put(c, map.getOrDefault(c,0)+1);
        
       	 int i=0, j=0;
        	 int count = map.size();
        	 int maxLength=Integer.MAX_VALUE, minStart=0, minEnd=0;
        	 while(j<s.length()){
           	     char c=s.charAt(j);
                     if(map.containsKey(c)){
                          int val = map.get(c)-1;
                          map.put(c, val);
    
                          if(val==0)
                             count--;
                      }
            
                      while(count==0){
                          if(maxLength>j-i+1){
                               maxLength = j-i+1;
                               minStart = i;
                               minEnd = j+1;
                          }
                
                          char x = s.charAt(i);
                          if(map.containsKey(x)){
                                map.put(x, map.get(x)+1);
                        
                                if(map.get(x)>0)
                                     count++;
                          }
                          i++;
                     }
                    j++;
              }
             return s.substring(minStart,minEnd);
        }
}